### PR TITLE
    Res validation should be aware of action prop reqs.

### PR DESCRIPTION
**Pull request by sabat**
Signed-off-by: Steve Abatangle <sabat@area51.org>

## Description
(Lamont has seen this patch before. It was never merged because of some confusion in the conversation, about a year ago. The bug remains in the most current Infra release.)

The `get` and `set` methods in the Property class (for resource validation) check that a property is required if its value is nil. However, they don't check to see whether the property was required for the *specific action* the resource is using, so any use of `get` and `set` will throw an exception if used with a resource that doesn't use a required property, even when the property is not required for that action.

What I'm describing is easier to understand with an example: the current version of the chef-client cookbook has a recipe (cron.rb) with a resource that uses the `:delete` action. When you try to `get` that resource, it throws an exception, explaining that `:command` is a required property—*but it's only required for the* `:create` *action, not for* `:delete`.

See https://github.com/chef-cookbooks/chef-client/blob/master/recipes/cron.rb#L95-L97

This PR fixes that by passing the resource's *action* to the `required?` method.

I have included a unit test update. I would also have included an integration test, but that seems to require deeper knowledge about Chef plumbing than I have.
## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [X] I have read the **CONTRIBUTING** document.
- [X] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [X] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: lib/chef/property.rb
Comment (jaymzh) [inline lib/chef/property.rb:427]: tests are failing because `resource.action` isn't defined...

File: lib/chef/property.rb
Comment (jaymzh) [inline lib/chef/property.rb:456]: same

**Review by jaymzh — CHANGES_REQUESTED** [review]
Logically this fix is good, but there's some bugs here, please get the tests passing.

File: lib/chef/property.rb
Comment (sabat) [reply lib/chef/property.rb:456]: Got it, hadn't noticed, will fix.

File: lib/chef/property.rb
Comment (marcparadise) [inline lib/chef/property.rb:308]: How does this functionally differ from the original check? When running through some scenarios, I can't come up with one that returns a different result given pre-conditions of  a non-nil action and :required having an Array.

Comment (sabat) [reply lib/chef/property.rb:308]: > How does this functionally differ from the original check?

The original cannot handle a check for multiple actions. It's fine if it's checking for just :create for example, but if `action` is an array with both :create and :delete it would fail with a simple `include?` test. I wrote this over a year ago and the details of the context now escape me, but IIRC there are situations where this needs to test for multiple actions—where `action` is going to be an array of actions, and `include?` only tests for the presence of a single object.

**Review by marcparadise — APPROVED** [review]

**Review by jaymzh — APPROVED** [review]
LGTM

**Comment by jaymzh** [comment]
I don't think the test failures are related, but we should get tests working before we merge this so we can see if this breaks anything - @marcparadise and @johnmccrae 

Also, @sabat - you have lint failuers.

Other than that seems good.

**Comment by sabat** [comment]
@jaymzh pushed a fix to that lint error—just a trailing space so easy to fix

**Comment by jaymzh** [comment]
Thanks!

Approved tests, lets see if we're in a better state

**Comment by jaymzh** [comment]
@sabat thanks! Can you rebase, all the test failures you are hitting have been fixed. Sorry!

**Comment by sabat** [comment]
@jaymzh rebased, pushed, build still failing in two Windows steps unrelated to this PR.

**Comment by tpowell-progress** [comment]
Ready to merge but holding off until we have Chef 18 RC packaged.

**Comment by tpowell-progress** [comment]
@sabat Now that GA has been released, can you please rebase so that we can confirm that tests still pass with these changes?

**Comment by johnmccrae** [comment]
@sabat one minor request - would you mind squashing your commits here to clean this up?

**Comment by johnmccrae** [comment]
I'll merge it after that.

**Comment by tpowell-progress** [comment]
@sabat Looks like you need to squash and rebase and verify all the commits have the DCO signed. Once that it done, we'll need to make sure the pipelines are still green for this PR.

**Comment by sabat** [comment]
> Looks like you need to squash and rebase and verify all the commits have
the DCO signed
Didn't I literally just do that? For a customer trying to help fix a flaw
in your product, this back-and-forth is getting frustrating.

**Comment by ramereth** [comment]
it looks like you had a lot of merged updates that caused your history to get all confusing. I went through and rebased it the best I could and pushed it here as one commit: https://github.com/ramereth/chef/tree/bugfix_rsrc_action_prop_validation

Let me know if that diff looks correct with all the fixes you expect.

**Pull request merged**
